lib/core/test/unit/Data/Map/NonEmpty/StrictSpec.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- Provides property tests for the strict implementation of a non-empty map.
+--
+-- In particular, we test that:
+--
+--   - each operation preserves the invariant.
+--
+--   - each operation behaves in exactly the same way as the equivalent
+--     operation provided by 'Data.Map.Strict'.
+--
+module Data.Map.NonEmpty.StrictSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Map.NonEmpty.Strict.Internal
+    ( NonEmptyMap, invariantHolds )
+import Test.Hspec
+    ( Spec, describe, it, parallel )
+import Test.Hspec.Core.QuickCheck
+    ( modifyMaxSuccess )
+import Test.QuickCheck
+    ( Arbitrary (..), Property, genericShrink, property, (===) )
+import Test.QuickCheck.Classes
+    ( eqLaws, foldableLaws, functorLaws, showLaws, traversableLaws )
+import Test.Utils.Laws
+    ( testLawsMany )
+
+import qualified Data.List.NonEmpty as NonEmptyList
+import qualified Data.Map.NonEmpty.Strict.Internal as NonEmptyMap
+import qualified Data.Map.Strict as Map
+
+spec :: Spec
+spec = describe "Data.Map.NonEmpty.Strict.NonEmptyMap" $
+
+    modifyMaxSuccess (const 1000) $ do
+
+        parallel $ describe "Class instances obey laws" $ do
+            testLawsMany @(NonEmptyMap Int Int)
+                [ eqLaws
+                , showLaws
+                ]
+            testLawsMany @(NonEmptyMap Int)
+                [ foldableLaws
+                , functorLaws
+                , traversableLaws
+                ]
+
+        parallel $ describe "Construction and deconstruction" $ do
+
+            it "prop_fromList_invariant" $
+                property prop_fromList_invariant
+            it "prop_fromList_toList" $
+                property prop_fromList_toList
+            it "prop_fromMap_invariant" $
+                property prop_fromMap_invariant
+            it "prop_fromMap_toMap" $
+                property prop_fromMap_toMap
+            it "prop_singleton_invariant" $
+                property prop_singleton_invariant
+            it "prop_singleton_toList" $
+                property prop_singleton_toList
+
+        parallel $ describe "Modification" $ do
+
+            it "prop_insert_invariant" $
+                property prop_insert_invariant
+            it "prop_insert" $
+                property prop_insert
+            it "prop_delete" $
+                property prop_delete
```

Comment:
Looks like we're missing a `prop_delete_invariant` here.

Reply:
It turns out we don't need an invariant test here, as the type of `delete` is `k -> NonEmptyMap k a -> Map k a`. The result is an ordinary map.